types/mithril/index.d.ts

```diff
@@ -208,7 +208,7 @@ declare namespace Mithril {
 	 * Any Javascript object that has a view method can be used as a Mithril component.
 	 * Components can be consumed via the m() utility.
 	 */
-	interface Component<Attrs, State extends Lifecycle<Attrs, State>> extends Lifecycle<Attrs, State> {
```

Comment:
Type inference isn't as frequently used in the tests (because they're testing that the types are correct), but it's used extensively by framework users. Our TS users tend to use type inference to a much greater degree than say Angular or Aurelia users, because our community as a whole prefers simple to verbose (like the framework itself).

The primary aim of the constraints is actually to enforce certain type invariants, which are likely not fully covered with tests:

1. `vnode.state === this` (i.e. `typeof vnode` &harr; `Vnode<Attrs, this>`) in all lifecycle methods.
1. For object components, `vnode.state` subtypes the component directly.
1. For class components, `vnode.state` is the class's instance (easy).
1. For closure (read: function) components, `vnode.state` is the function's return value.

This is really why the constraints are used, and why I said `State extends this` would be optimal for object components. If `ThisType<T>` can help with this, then that's what we need.

Sorry if I phrased it poorly and incorrectly the first time.